src/Microsoft.SqlTools.ServiceLayer/Cms/CmsService.cs

```diff
@@ -0,0 +1,434 @@
+            ServerConnection serverConn = ConnectionService.OpenServerConnection(connectionInfo);
+
+            return serverConn;
+        }
+
+        private ServerConnection ValidateAndCreateConnection(string ownerUri)
+        {
+            ServerConnection serverConn = null;
+            if (ownerUri != null)
+            {
+                ConnectionInfo connInfo = null;
+                if (ConnectionServiceInstance.TryFindConnection(ownerUri, out connInfo))
+                {
+                    serverConn = ConnectionService.OpenServerConnection(connInfo);
+                }
+            }
+            return serverConn;
+        }
+
+        private ListRegisteredServersResult GetChildrenfromParentGroup(ServerGroup parentGroup)
+        {
+            var servers = parentGroup.RegisteredServers;
+            var groups = parentGroup.ServerGroups;
+
+            // Convert to appropriate variables and return
+            var serverResults = new List<RegisteredServerResult>();
+            foreach (var s in servers)
+            {
+                serverResults.Add(new RegisteredServerResult
+                {
+                    Name = s.Name,
+                    ServerName = s.ServerName,
+                    Description = s.Description,
+                    connectionDetails = ConnectionServiceInstance.ParseConnectionString(s.ConnectionString),
+                    RelativePath = s.KeyChain.Urn.SafeToString()
+                });
+            }
+
+            var groupsResults = new List<RegisteredServerGroup>();
+            foreach (var s in groups)
+            {
+                groupsResults.Add(new RegisteredServerGroup
+                {
+                    Name = s.Name,
+                    Description = s.Description,
+                    RelativePath = s.KeyChain.Urn.SafeToString()
+                });
+            }
+            ListRegisteredServersResult result = new ListRegisteredServersResult() { RegisteredServersList = serverResults, RegisteredServerGroups = groupsResults };
+            return result;
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Internal for testing purposes only
+        /// </summary>
+        internal static ConnectionService ConnectionServiceInstance
+        {
+            get
+            {
+                if (connectionService == null)
+                {
+                    connectionService = ConnectionService.Instance;
+                }
+                return connectionService;
+            }
+            set
+            {
+                connectionService = value;
+            }
+        }
+
+        /// <summary>
+        /// Internal for testing purposes only
+        /// </summary>
+        internal static RegisteredServersStore ServerStore
+        {
+            get
+            {
+                if (registerdServerStore == null)
+                {
+                    registerdServerStore = RegisteredServersStore.LocalFileStore;
```

Comment:
should you assign using the setter (i.e. `this.ServerStore = …`) to go through the lock, or take the lock explicitly here?